adam/learner/attributes.py

```diff
@@ -297,3 +304,192 @@ def _update_hypothesis(
+        # since relations, verbs, etc. build on object semantic nodes.
+        # Furthermore we only need one node, because attributes only involve one object.
+        return (
+            len(
+                language_perception_semantic_alignment.perception_semantic_alignment.semantic_nodes
+            )
+            > 1
+        )
+
+    def _preprocess_scene(
+        self, perception_semantic_alignment: PerceptionSemanticAlignment
+    ) -> PerceptionSemanticAlignment:
+        return perception_semantic_alignment
+
+    def log_hypotheses(self, log_output_path: Path) -> None:
+        logging.info(
+            "Logging %s hypotheses to %s",
+            len(self._concept_to_hypotheses_and_scores),
+            log_output_path,
+        )
+        for (
+            concept,
+            hypotheses_to_scores,
+        ) in self._concept_to_hypotheses_and_scores.items():
+            for (i, (hypothesis, score)) in enumerate(hypotheses_to_scores.items()):
+                hypothesis.render_to_file(
+                    concept.debug_string,
+                    log_output_path / f"{concept.debug_string}.{i}.score_{score}",
+                )
+
+        logging.info(
+            "Logging %s lexicon items to %s", len(self._lexicon), log_output_path
+        )
+        for (
+            surface_template,
+            meaning_as_perception_graph_template,
+        ) in self._lexicon.items():
+            meaning_as_perception_graph_template.render_to_file(
+                surface_template.to_short_string(),
+                log_output_path / f"lexicon-{surface_template.to_short_string()}",
+            )
+
+
+def _extract_candidate_attributes(
+    whole_scene_perception_graph: PerceptionGraph,
+    object_with_attribute: ObjectSemanticNode,
+) -> Sequence[PerceptionGraph]:
+    perception_digraph = whole_scene_perception_graph.copy_as_digraph()
+    # For now, we assume all attributes are based on properties.
+    properties = immutableset(
+        [
+            node
+            for _, node, label in perception_digraph.out_edges(
+                object_with_attribute, data="label"
+            )
+            if label == HAS_PROPERTY_LABEL
+        ]
+    )
+    # candidate_attribute_subgraph_node_sets = [
+    #     immutableset(chain([object_with_attribute], properties_subset))
+    #     for properties_subset in powerset(properties)
+    #     if properties_subset
+    # ]
+    # For testing purposes, use the much smaller set of candidate attributes
+    # where we consider only single properties as possible attributes.
+    candidate_attribute_subgraph_node_sets = [
+        immutableset([object_with_attribute, property]) for property in properties
+    ]
+    return immutableset(
+        [
+            whole_scene_perception_graph.subgraph_by_nodes(
+                candidate_attribute_subgraph_nodes
+            )
+            for candidate_attribute_subgraph_nodes in candidate_attribute_subgraph_node_sets
+        ]
+    )
```

Comment:
This is the section of code we need to be sure to fully disclaim in any writeup which uses this learner. We know that this is a 'hack' to substantially cut down on the search space.

Reply:
@LichMaster98 So just to confirm, do we want to keep this hack then?

Reply:
Yes @spigo900 otherwise the search space for pursuit could grow too large.

Reply:
@LichMaster98 Definitely, just thought I'd confirm. Thanks!